src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Debugger_InProc.cs

```diff
@@ -33,5 +56,105 @@ public void SetBreakPoint(string fileName, int lineNumber, int columnIndex)
+                    }
+
+                    actionSucceeded = false;
+                }
+
+                return actionSucceeded;
+            };
+
+            // Repeat the the command if "Operation Not Supported" is thrown.
+            if (!this.TryWaitFor(DebuggerCommandRetryTimeout, predict))
+            {
+                string message = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "Failed to raise debugger command within '{0}' seconds.",
+                    DebuggerCommandRetryTimeout.TotalSeconds);
+
+                throw new Exception(message);
+            }
+        }
+
+        /// <summary>
+        /// Polls for the specified delegate to return true for the given timeout.
+        /// </summary>
+        /// <param name="timeout">Timeout to keep polling.</param>
+        /// <param name="predicate">Delegate to invoke.</param>
+        /// <returns>True if the delegate returned true when polled, otherwise false.</returns>
+        public bool TryWaitFor(TimeSpan timeout, Func<bool> predicate)
+        {
+            return this.TryWaitFor(timeout, DefaultPollingInterCallSleep, predicate);
+        }
+
+        /// <summary>
+        /// Polls for the specified delegate to return true for the given timeout.
+        /// </summary>
+        /// <param name="timeout">Timeout to keep polling.</param>
+        /// <param name="interval">Time to wait between polling.</param>
+        /// <param name="predicate">Delegate to invoke.</param>
+        /// <returns>
+        /// True if the delegate returned true when polled, otherwise false.
+        /// </returns>
+        private bool TryWaitFor(TimeSpan timeout, TimeSpan interval, Func<bool> predicate)
+        {
+            DateTime endTime = DateTime.UtcNow + this.ApplyTimeoutMultiplier(timeout);
+            bool validationDelegateSuccess = false;
+
+            while (DateTime.UtcNow < endTime)
+            {
+                // Note: we don't do this inline in the while() condition and return the result of 
+                // (DateTime.Now < startTime + timeout) as this could lead to cases where the validation
+                // delegate returned true, at the boundary of the valid time, and would return false
+                // when hitting the return statement. 
+                if (predicate.Invoke())
+                {
+                    validationDelegateSuccess = true;
+                    break;
+                }
+
+                System.Threading.Thread.Sleep(interval);
+            }
+
+            return validationDelegateSuccess;
+        }
+
+
+        /// <summary>
+        /// Multiplies the given TimeSpan by the current global timeout multiplier.
+        /// </summary>
+        /// <param name="timeSpan">A timeout value to multiply</param>
+        /// <returns>The product of the given timespan and the global timeout multiplier</returns>
+        private TimeSpan ApplyTimeoutMultiplier(TimeSpan timeSpan)
```

Comment:
>TimeSpan [](start = 16, length = 8)

All these helpers can be static

Reply:
agree. thanks!

---
In reply to: [136209233](https://github.com/dotnet/roslyn/pull/21824#discussion_r136209233) [](ancestors = 136209233)